pkg/kubelet/pod_workers.go

```diff
@@ -696,11 +710,70 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
 	return gracePeriod, status.gracePeriod != 0 && status.gracePeriod != gracePeriod
 }
 
+// isStaticPodAllowedToStart returns true if there is no other started static
+// pod with the same full name and pod is the oldest static pod waiting to
+// start.
+func (p *podWorkers) isStaticPodAllowedToStart(pod *v1.Pod) bool {
+	fullname := kubecontainer.GetPodFullName(pod)
+	uid, started := p.startedStaticPodsByFullname[fullname]
+	if started {
+		if uid == pod.UID {
+			// pod is already started
+			return true
+		}
+		return false
+	}
+
+	waitingToStartPods := p.waitingToStartStaticPodsByFullname[fullname]
+	if len(waitingToStartPods) == 0 {
+		return false
+	}
+
+	return pod.UID == waitingToStartPods[0]
+}
+
+// startNextStaticPod starts next static pod for the given fullname
+func (p *podWorkers) startNextStaticPod(fullname string) {
+	waitingToStartPods := p.waitingToStartStaticPodsByFullname[fullname]
+	if len(waitingToStartPods) == 0 {
+		return
+	}
+
+	p.startedStaticPodsByFullname[fullname] = waitingToStartPods[0]
+	p.waitingToStartStaticPodsByFullname[fullname] = waitingToStartPods[1:]
+	if len(p.waitingToStartStaticPodsByFullname[fullname]) == 0 {
+		delete(p.waitingToStartStaticPodsByFullname, fullname)
+	}
+}
+
 func (p *podWorkers) managePodLoop(podUpdates <-chan podWork) {
 	var lastSyncTime time.Time
+
+	var staticPodStarted bool
 	for update := range podUpdates {
 		pod := update.Options.Pod
 
+		if kubetypes.IsStaticPod(pod) {
+			if !staticPodStarted {
+				// try to start static pod
+				func() {
+					p.podLock.Lock()
```

Comment:
This needs to be a function, and IsStaticPod(pod) should be inside that function.

```
if p.allowPodStart(pod) {
```

```
func allowPodStart(pod) bool {
  if !kubetypes.IsStaticPod(pod) { 
     return true
  }
  ...
  p.podLock.Lock()
  ...
}
```

Reply:
In general "if static pod" is an assumption that will change, so main code loops should be shielded from knowing about that.